fix issue #20922 by handling missing expr in `exprList` for `tkOf`

Vindaar opened the pull request:
Not quite sure if this is the right way to fix #20922 especially because `nim check` still happily prints the illformed AST message in addition (but that may well be expected behavior?). Anyhow, as this was simple enough to do, I'll let you decide. :)

Vindaar committed:
fix issue #20922 by handling missing expr in `exprList` for `tkOf`

metagn commented:
The grammar for `exprList` uses `^+`, which means *all* exprLists should have 1 or more expressions. Are there any exceptions to this in reality?

metagn commented:
Apparently only `except` uses `exprList` while logically allowing 0 expressions. We can make a copy of the old proc with the name `optionalExprList` and use it for `except`, then for `exprList` we can do:

```diff
+var a = parseExpr(p)
+result.add(a)
  while (p.tok.tokType != endTok) and (p.tok.tokType != tkEof):
-   var a = parseExpr(p)
-   result.add(a)
    if p.tok.tokType != tkComma: break
    getTok(p)
    optInd(p, a)
+   a = parseExpr(p)
+   result.add(a)
```

Vindaar committed:
fix line numbers in test case

Vindaar commented:
On the one hand that cleans up my patch, but on the other it duplicates more code. Not sure what we prefer?

metagn commented:
Yeah, maybe we could just add an `optional: bool = false` parameter (only true for `except`) and track if an expression was parsed. The grammar would still need an `optionalExprList = expr ^* comma` though (again, only used by `except`).


```nim
  var exprParsed = false
  while (p.tok.tokType != endTok) and (p.tok.tokType != tkEof):
   var a = parseExpr(p)
   exprParsed = true
   result.add(a)
    if p.tok.tokType != tkComma: break
    getTok(p)
    optInd(p, a)
  if not optional and not exprParsed:
    error("expression expected")
```

Araq commented:
You need to do what @metagn said. I prefer code duplication over flags as I cannot use "flags" in the grammar and the parser should reflect the grammar as closely as possible.

Vindaar committed:
rewrite exprList requiring expr, add optionalExprList for except

As suggested by @metagn.

Vindaar committed:
update test case to reflect new code

Vindaar commented:
Ok, just updated the code accordingly. The resulting error message is more cryptic again now of course, but I guess it's in line with what one would expect otherwise. The error messages for the test case using `nim check` is even more scary now though, haha.

I hope I updated the grammar rules correctly. I understand correctly that the `doc/grammar.txt` file is somehow generated from the comments in `parser.nim`?

Vindaar commented:
Huh, guess we found a bug in existing code: https://github.com/jangko/nim-noise/blob/master/noise/editorImpl.nim#L202 ?

metagn commented:
Come to think of it, the `optional = false` method should be fine if we just use `exprList?`.

To fix CI: 1. `nim r --lib:lib compiler/parser`, 2. fix the line numbers in the test, or just use [tt.Error](https://github.com/nim-lang/Nim/blob/10b6e4f5b4a119a8fa42f82a24ae7e4b08b2c103/tests/specialops/terrmsgs.nim#L20)

Vindaar committed:
update grammar.txt

Vindaar committed:
update line numbers in test case taking into account nimout

Given the number of errors that are produced it seems easier to do it
this way instead of using `tt.Error`.

Vindaar commented:
> Come to think of it, the `optional = false` method should be fine if we just use `exprList?`.
> 
> To fix CI: 1. `nim r --lib:lib compiler/parser`, 2. fix the line numbers in the test, or just use [tt.Error](https://github.com/nim-lang/Nim/blob/10b6e4f5b4a119a8fa42f82a24ae7e4b08b2c103/tests/specialops/terrmsgs.nim#L20)

Thanks a lot for the guidance!

I decided to update the output instead of using `tt.Error` as it seems like that would be difficult with the number of errors produced.

On the context of changing to an `optional = false` flag instead, I don't have an opinion. So you say in that case we don't need to add `optionalExprList`? Would the grammar have to change from 
```
  #| exprList = expr ^+ comma
# to
  #| exprList = expr ^* comma
```
then?

metagn commented:
No, `exprList = expr ^+ comma` would stay, we would just use `'except' exprList?` instead of `'except' exprList` in the grammar.

Don't want to backseat though, this is probably good enough.

Araq commented:
Nice job, but breaks a package:


  /home/runner/.nimble/pkgs/noise-0.1.6/noise/editorImpl.nim(202, 40) Error: expression expected, but found ':'
  stack trace: (most recent call last)
  /tmp/nimblecache-946033505/nimscriptapi_674286946.nim(187, 16)
  /home/runner/work/Nim/Nim/pkgstemp/inim/inim.nimble(20, 8) testTask
  /home/runner/work/Nim/Nim/lib/system/nimscript.nim(264, 7) exec
  /home/runner/work/Nim/Nim/lib/system/nimscript.nim(264, 7) Error: unhandled exception: FAILED: nim c -d:NoColor -d:prompt_no_history --out:bin/inim inim.nim [OSError]
       Error: Exception raised during nimble script execution

Vindaar commented:
> Nice job, but breaks a package:

Yup, as mentioned further up. Already did a PR. @jangko do you mind tagging a new version for the CI to pass?

ringabout commented:
@Vindaar You can also patch testament like `pkg "inim", "nimble install noise@#head; nimble test"`.

metagn commented:
Maybe support trailing commas?

jangko commented:
done
